### PR TITLE
feat(module:upload): add nz-upload component

### PR DESCRIPTION
**Pull request by cipchk**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/showcase/nz-demo-upload/nz-demo-upload-avatar.component.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  }
+  `]
+})
+export class NzDemoUploadAvatarComponent {
+  loading = false;
+  avatarUrl: string;
+
+  constructor(private msg: NzMessageService) {}
+
+  beforeUpload(file: File): boolean {
```
Comment (vthinkxie) [inline src/showcase/nz-demo-upload/nz-demo-upload-avatar.component.ts]: should be an arrow function here.

File: src/showcase/nz-demo-upload/nz-demo-upload-basic.component.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'nz-demo-upload-basic',
+  template: `
+  <nz-upload
+    nzAction="https://jsonplaceholder.typicode.com/posts/">
+    <button nz-button>
+      <span><i class="anticon anticon-upload"></i>Click to Upload</span>
```
Comment (vthinkxie) [inline src/showcase/nz-demo-upload/nz-demo-upload-basic.component.ts]: should be <i class="anticon anticon-upload"></i> and <span>Click to Upload</span>

File: src/showcase/nz-demo-upload/nz-demo-upload-manually.component.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { HttpRequest, HttpClient, HttpResponse } from '@angular/common/http';
+import { NzMessageService } from './../../components/message/nz-message.service';
+import { UploadFile } from './../../components/upload/interface';
+
+@Component({
+  selector: 'nz-demo-upload-manually',
+  template: `
+  <nz-upload
+    [(nzFileList)]="fileList"
+    [nzBeforeUpload]="beforeUpload.bind(this)">
```
Comment (vthinkxie) [inline src/showcase/nz-demo-upload/nz-demo-upload-manually.component.ts]: why need bind here?

File: src/components/upload/nz-upload.component.ts [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+@Component({
+  selector: 'nz-upload',
+  template: `
+  <ng-template #list>
+    <nz-upload-list *ngIf="nzShowUploadList"
+      [listType]="nzListType"
+      [items]="nzFileList"
+      [icons]="nzShowUploadList"
+      [onPreview]="nzPreview"
+      [onRemove]="onRemove.bind(this)"></nz-upload-list>
```
Comment (vthinkxie) [inline src/components/upload/nz-upload.component.ts]: onRemove be an arrow function other than bind here.

**Comment by wilsoncook** [comment]
@cipchk Demo needs to be corrected:
image

**Comment by wilsoncook** [comment]
The `msg` may be missed:
image

**Comment by wilsoncook** [comment]
@cipchk Should be a way to catch errors if not success? And when failed, the loading icon below seems stucked.
image

**Comment by cipchk** [comment]
除demo样式以外，所有都是由于未统一箭头函数引起的 this 对象问题，对此同时增加文档描述。

**Comment by MR-ZiWeiter** [comment]
cli/1.65 版本，使用upload组件似乎不行
image

**Comment by cipchk** [comment]
`OPTIONS` 请求失败，这大部分情况是来自你后端的因素。

**Pull request merged**